codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +66,825 @@
-        if (executor.inEventLoop()) {
-            writeFromStreamChannel0(msg, flush, promise);
-        } else {
-            try {
-                executor.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        writeFromStreamChannel0(msg, flush, promise);
-                    }
-                });
-            } catch (Throwable cause) {
-                promise.setFailure(cause);
+    @Override
+    final void onHttp2FrameStreamException(ChannelHandlerContext ctx, Http2FrameStreamException cause) {
+        Http2FrameStream stream = cause.stream();
+        DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
+
+        if (childChannel == null)  {
+            // TODO: Should we log this ?
+            return;
+        }
+        try {
+            childChannel.pipeline().fireExceptionCaught(cause.getCause());
+        } finally {
+            childChannel.close();
+        }
+    }
+
+    private void onHttp2StreamFrame(DefaultHttp2StreamChannel childChannel, Http2StreamFrame frame) {
+        if (!childChannel.fireChildRead(frame)) {
+            if (childChannel.fireChildReadComplete()) {
+                flushNeeded = true;
+            }
+
+            // Just called fireChildReadComplete() no need to do again in channelReadComplete(...)
+            childChannel.fireChannelReadPending = false;
+        } else if (!childChannel.fireChannelReadPending) {
+
+            assert childChannel.next == null;
+
+            if (tail == null) {
+                assert head == null;
+                tail = head = childChannel;
+            } else {
+                tail.next = childChannel;
+                tail = childChannel;
             }
+            childChannel.fireChannelReadPending = true;
         }
     }
 
-    private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise promise) {
+    private void onHttp2GoAwayFrame(ChannelHandlerContext ctx, final Http2GoAwayFrame goAwayFrame) {
         try {
-            write(ctx, msg, promise);
-        } catch (Throwable cause) {
-            promise.tryFailure(cause);
-        }
-        if (flush) {
-            flush(ctx);
+            forEachActiveStream(new Http2FrameStreamVisitor() {
+                @Override
+                public boolean visit(Http2FrameStream stream) {
+                    final int streamId = stream.id();
+                    final DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
+                    if (streamId > goAwayFrame.lastStreamId() && isOutboundStream(connection().isServer(), streamId)) {
+                        childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+                    }
+                    return true;
+                }
+            });
+        } catch (Http2Exception e) {
+            // TODO: log me ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)